File: matlab/vals2colors.m
function[colors] = vals2colors(vals, cmap, res)
%VALS2COLORS project a vector of scalar values into (R,G,B) space
%
% Usage: colors = vals2colors(vals, [cmap], [res]);
%
% INPUTS:
%   vals: a vector of scalar values.  If vals is a matrix, it will be
%         re-shaped into a column vector.
%
%   cmap: an optional argument specifying (in a character string) which
%         colormap to use.  Default: 'linspecer'.
%
%    res: an optional argument specifying the maximum number of unique
%         colors to use.  Default: 100.
% 
% OUTPUTS:
% colors: a numel(vals) by 3 matrix of colors.
%
% SEE ALSO: PLOT_COORDS, LINSPECER
%
%  AUTHOR: Jeremy R. Manning
% CONTACT: jeremy.r.manning@dartmouth.edu

% CHANGELOG:
% 4-18-16  jrm  wrote it.


%convert values to RGB values using the specified colormap and resolution
if ~exist('res', 'var'), res = 100; end
if ~exist('cmap', 'var'), cmap = 'linspecer'; end

cmap = eval(sprintf('%s(%d)', cmap, round(res)));

vals = vals(:);
ranks = arrayfun(@(i)(sum(vals <= i)), vals);
ranks = round(res.*ranks./length(vals));
ranks(ranks < 1) = 1;

colors = cmap(ranks, :);

File: python/vals2colors.py
import seaborn as sns
import itertools

def vals2colors(vals,cmap='husl',res=100):
	"""Maps values to colors
	Args:
	values (list or list of lists) - list of values to map to colors
	cmap (str) - color map (default is 'husl')
	Returns:
	list of rgb tuples
	"""
	# flatten if list of lists
	if any(isinstance(el, list) for el in vals):
		vals = list(itertools.chain(*vals))

	# get palette from seaborn
	palette = sns.color_palette(cmap, res)

	# rank the values and then normalize
	ranks = list(map(lambda x: sum([val <= x for val in vals]),vals))
	ranks = list(map(lambda rank: int(round(res*rank/len(vals))),ranks))

	return [palette[rank-1] for rank in ranks]
